test/spinxsk/spinxsk.c

```diff
@@ -1992,6 +2022,182 @@ ProcessPkts(
+            UINT8 destConnId[20];
+            FillRandomBuffer(sizeof(destConnId), destConnId);
+            UINT8 srcConnIdLength = (UINT8)(RandUlong() % (XDP_QUIC_MAX_CID_LENGTH + 1));
+            UINT8 srcConnId[20];
+            FillRandomBuffer(sizeof(srcConnId), srcConnId);
+            BOOLEAN useShortHeader = (BOOLEAN)(RandUlong() % 2);
+            ASSERT_FRE(PktBuildQuicPacket(
+                payload, &payloadLength, frame, quicPayloadLength, typeSpecificBits, version,
+                destConnId, destConnIdLength, srcConnId, srcConnIdLength, useShortHeader));
+        } else {
+            //
+            // Use a random payload.
+            //
+            payloadLength = RandUlong() % (payloadBufferSize + 1);
+        }
+
+        //
+        // Build a transport frame
+        //
+        const UINT32 backfill = RandUlong() % (FNMP_FRAME_MAX_BACKFILL + 1);
+        const UINT16 localPort = (UINT16)RandUlong();
+        const UINT16 remotePort = (UINT16)RandUlong();
+
+        UINT32 frameLength = frameBufferSize - backfill;
+        UINT32 frameType = RandUlong() % 3;
+        if (frameType == 0) {
+            //
+            // Build a UDP frame.
+            //
+            ASSERT_FRE(PktBuildUdpFrame(
+                &frame[backfill], &frameLength, payload, (UINT16)payloadLength,
+                &localMac, &remoteMac, af, &localIp, &remoteIp, localPort, remotePort));
+        } else if (frameType == 1) {
+            //
+            // Build a TCP frame.
+            // Up to 40 bytes of TCP options and random flags.
+            //
+            UINT8 tcpOptions[TCP_MAX_OPTION_LEN];
+            FillRandomBuffer(sizeof(tcpOptions), tcpOptions);
+            UINT16 tcpOptionsLength = 4 * (RandUlong() % (10 + 1));
+            const UINT32 thSeq = RandUlong();
+            const UINT32 thAck = RandUlong();
+            const UINT8 thFlags = (UINT8)RandUlong();
+            const UINT16 thWin = (UINT16)RandUlong();
+
+            ASSERT_FRE(PktBuildTcpFrame(
+                &frame[backfill], &frameLength, payload, (UINT16)payloadLength,
+                tcpOptions, tcpOptionsLength, thSeq, thAck, thFlags, thWin,
+                &localMac, &remoteMac, af, &localIp, &remoteIp, localPort, remotePort));
+        } else {
+            //
+            // Send pure chaos.
+            //
+            frameLength = RandUlong() % (frameLength + 1);
+            for (UINT32 j = 0; j + sizeof(ULONG) < frameLength; j += sizeof(ULONG)) {
+                *(ULONG*)(frame + backfill + j) = RandUlong();
+            }
+        }
+
+        //
+        // Setup the fnmp data frame.
+        //
+        DATA_FRAME fnmpFrame = {0};
+        fnmpFrame.Input.RssHashQueueId = RandUlong() % queueCount;
+        fnmpFrame.Input.Checksum.Value = (PVOID)(ULONG_PTR)RandUlong();
+        fnmpFrame.Input.Rsc.Value = (PVOID)(ULONG_PTR)RandUlong();
+
+        //
+        // Split the frame into one or two MDLs.
```

Comment:
i think we need at minimum three MDLs to cover every possible weird case, aside from the very-weird 64+ MDL case that XDP has a [special case](https://github.com/microsoft/xdp-for-windows/blob/861c620a2136388e2bb1abc239b28b76e538b00d/src/xdplwf/recv.c#L1437C1-L1462) for.

Reply:
Adding a 3 MDL.